diff -Nur oscam-svn-modern-zeroconf-emu-icam/module-dvbapi.c oscam-svn-modern-zeroconf-emu-icam-sogno/module-dvbapi.c
--- oscam-svn-modern-zeroconf-emu-icam/module-dvbapi.c	2022-12-30 19:23:46.317945224 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno/module-dvbapi.c	2022-12-30 19:25:43.415112564 +0100
@@ -352,7 +352,7 @@
 struct s_client *dvbapi_client;
 
 const char *boxdesc[] = { "none", "dreambox", "duckbox", "ufs910", "dbox2", "ipbox", "ipbox-pmt",
-							"dm7000", "qboxhd", "coolstream", "neumo", "pc", "pc-nodmx", "samygo" };
+							"dm7000", "qboxhd", "coolstream", "neumo", "pc", "pc-nodmx", "samygo", "sogno" };
 
 // when updating devices[BOX_COUNT] make sure to update these index defines
 #define BOX_INDEX_QBOXHD 0
@@ -1718,12 +1718,19 @@
 
 							if(!match)
 							{
-								for(j = 0; j < CA_MAX; j++)
+								if (cfg.dvbapi_boxtype == BOXTYPE_SOGNO)
+								{								
+									remove_streampid_from_list(demux[demux_id].ca_mask, pidtobestopped, idx);
+								}								
+								else
 								{
-									if(((demux[demux_id].ca_mask & (1 << j)) == (uint32_t) (1 << j)))
+									for(j = 0; j < CA_MAX; j++)
 									{
-										remove_streampid_from_list(j, pidtobestopped, idx);
-										break;
+										if(((demux[demux_id].ca_mask & (1 << j)) == (uint32_t) (1 << j)))
+										{
+											remove_streampid_from_list(j, pidtobestopped, idx);
+											break;
+										}
 									}
 								}
 							}
@@ -2387,46 +2394,92 @@
 						action != FOUND_STREAMPID_INDEX && action != ADDED_STREAMPID_INDEX &&
 						action != REMOVED_STREAMPID_INDEX)
 					{
-						// removed last index of this streampid on ca? -> disable this pid with -1 on this ca
-						if((action == REMOVED_STREAMPID_LASTINDEX || action == FIRST_STREAMPID_INDEX)
-							&& (is_ca_used(i, streampid) == INDEX_INVALID))
+						// removed last index of this streampid on ca? -> disable this pid with -1 on this ca						
+						if (cfg.dvbapi_boxtype == BOXTYPE_SOGNO)
+						{	
+							if((action == REMOVED_STREAMPID_LASTINDEX)
+								&& (is_ca_used(i, streampid) == INDEX_INVALID))
+							{
+								curidx = DVBAPI_INDEX_DISABLE;
+							}
+						}
+						else
 						{
-							curidx = DVBAPI_INDEX_DISABLE;
+							if((action == REMOVED_STREAMPID_LASTINDEX || action == FIRST_STREAMPID_INDEX)
+								&& (is_ca_used(i, streampid) == INDEX_INVALID))
+							{
+								curidx = DVBAPI_INDEX_DISABLE;
+							}
 						}
-
 						// removed index of streampid that is used to decode on ca -> get a fresh one
-						if(action == REMOVED_DECODING_STREAMPID_INDEX || action == FIRST_STREAMPID_INDEX)
+						
+						if (cfg.dvbapi_boxtype == BOXTYPE_SOGNO)
+						{	
+							if(action == REMOVED_DECODING_STREAMPID_INDEX)
+							{
+								newidx = is_ca_used(i, streampid); // get an active index for this pid and enable it on ca device
+								curidx = DVBAPI_INDEX_DISABLE;
+							}
+						}
+						else
 						{
-							newidx = is_ca_used(i, streampid); // get an active index for this pid and enable it on ca device
-							curidx = DVBAPI_INDEX_DISABLE;
+							if(action == REMOVED_DECODING_STREAMPID_INDEX || action == FIRST_STREAMPID_INDEX)
+							{
+								newidx = is_ca_used(i, streampid); // get an active index for this pid and enable it on ca device
+								curidx = DVBAPI_INDEX_DISABLE;
+							}
 						}
-
 						while(curidx != INDEX_INVALID || newidx != INDEX_INVALID)
 						{
 							memset(&ca_pid2, 0, sizeof(ca_pid2));
 							ca_pid2.pid = streampid;
 							if(curidx != INDEX_INVALID)
 							{
-								(curidx == DVBAPI_INDEX_DISABLE) ? (ca_pid2.index = -1) : (ca_pid2.index = curidx);
-								cs_log_dbg(D_DVBAPI, "Demuxer %d %s stream %d pid=0x%04x index=%d on ca%d",
-										demux_id,
-										((enable && curidx != DVBAPI_INDEX_DISABLE) ? "enable" : "disable"),
-										num + 1,
-										ca_pid2.pid,
-										ca_pid2.index,
-										i);
+								(curidx == DVBAPI_INDEX_DISABLE) ? (ca_pid2.index = -1) : (ca_pid2.index = curidx);								
+								if (cfg.dvbapi_boxtype == BOXTYPE_SOGNO)
+								{
+									cs_log_dbg(D_DVBAPI, "Demuxer %d %s stream %d pid=0x%04x index=%d on ca%d",
+											demux_id,
+											(enable ? "enable" : "disable"),
+											num + 1,
+											ca_pid2.pid,
+											ca_pid2.index,
+											i);
+								}
+								else
+								{
+									cs_log_dbg(D_DVBAPI, "Demuxer %d %s stream %d pid=0x%04x index=%d on ca%d",
+											demux_id,
+											((enable && curidx != DVBAPI_INDEX_DISABLE) ? "enable" : "disable"),
+											num + 1,
+											ca_pid2.pid,
+											ca_pid2.index,
+											i);
+								}
 								curidx = INDEX_INVALID; // flag this index as handled
 							}
 							else if(newidx != INDEX_INVALID)
 							{
 								(newidx == DVBAPI_INDEX_DISABLE) ? (ca_pid2.index = -1) : (ca_pid2.index = newidx);
-								cs_log_dbg(D_DVBAPI, "Demuxer %d %s stream %d pid=0x%04x by index=%d on ca%d",
-										demux_id,
-										((enable && action == FIRST_STREAMPID_INDEX) ? "enable" : "takeover"),
-										num + 1,
-										ca_pid2.pid,
-										ca_pid2.index,
-										i);
+								if (cfg.dvbapi_boxtype == BOXTYPE_SOGNO)
+								{
+									cs_log_dbg(D_DVBAPI, "Demuxer %d takeover stream %d pid=0x%04x by index=%d on ca%d",
+											demux_id,
+											num + 1,
+											ca_pid2.pid,
+											ca_pid2.index,
+											i);
+								}
+								else
+								{
+									cs_log_dbg(D_DVBAPI, "Demuxer %d %s stream %d pid=0x%04x by index=%d on ca%d",
+											demux_id,
+											((enable && action == FIRST_STREAMPID_INDEX) ? "enable" : "takeover"),
+											num + 1,
+											ca_pid2.pid,
+											ca_pid2.index,
+											i);
+								}
 								newidx = INDEX_INVALID; // flag this takeover / new index as handled
 							}
 
diff -Nur oscam-svn-modern-zeroconf-emu-icam/module-dvbapi.h oscam-svn-modern-zeroconf-emu-icam-sogno/module-dvbapi.h
--- oscam-svn-modern-zeroconf-emu-icam/module-dvbapi.h	2022-12-30 18:47:13.382386277 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno/module-dvbapi.h	2022-12-30 19:25:43.415112564 +0100
@@ -42,7 +42,8 @@
 #define BOXTYPE_PC         11
 #define BOXTYPE_PC_NODMX   12
 #define BOXTYPE_SAMYGO     13
-#define BOXTYPES           13
+#define BOXTYPE_SOGNO      14
+#define BOXTYPES           14
 #define DMXMD5HASHSIZE     16 // use MD5()
 
 // we store the results of remove_streampid_from_list()
