diff -Nur oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/CMakeLists.txt oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/CMakeLists.txt
--- oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/CMakeLists.txt	2022-12-30 19:23:46.313945320 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/CMakeLists.txt	2022-12-31 13:44:02.282047174 +0100
@@ -666,9 +666,15 @@
 	touch ${CMAKE_CURRENT_SOURCE_DIR}/oscam-log.c
 )
 
-execute_process (COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/config.sh --oscam-revision COMMAND tr -d '\n' OUTPUT_VARIABLE CS_SVN_VERSION)
+execute_process (COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/config.sh --oscam-revision COMMAND cut -d "+" -f4 COMMAND tr -d '\n' OUTPUT_VARIABLE CS_SVN_VERSION)
 add_definitions ("-D'CS_SVN_VERSION=\"${CS_SVN_VERSION}\"'")
 
+execute_process (COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/config.sh --oscam-revision COMMAND cut -d "+" -f1-3 COMMAND tr -d '\n' OUTPUT_VARIABLE CS_SMOD_VERSION)
+add_definitions ("-D'CS_SMOD_VERSION=\"${CS_SMOD_VERSION}\"'")
+
+execute_process (COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/config.sh --oscam-revision COMMAND cut -d "+" -f3 COMMAND tr -d '\n' OUTPUT_VARIABLE CS_SMOD_VERSION_HASH)
+add_definitions ("-D'CS_SMOD_VERSION_HASH=\"${CS_SMOD_VERSION_HASH}\"'")
+
 execute_process(COMMAND ${CMAKE_C_COMPILER} -dumpmachine COMMAND tr -d '\n' OUTPUT_VARIABLE CS_TARGET)
 add_definitions ("-D'CS_TARGET=\"${CS_TARGET}\"'")
 #----------------------- global compile and link options ------------------------------
diff -Nur oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/config.h oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/config.h
--- oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/config.h	2022-12-30 19:23:46.313945320 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/config.h	2022-12-31 13:46:06.471044138 +0100
@@ -7,25 +7,25 @@
 #define WEBIF 1
 #define WEBIF_LIVELOG 1
 #define WEBIF_JQUERY 1
-//#define TOUCH 1
+#define TOUCH 1
 #define WITH_SSL 1
 #if defined(__linux__) || defined(__CYGWIN__)
 #define HAVE_DVBAPI 1
 #endif
 //#define WITH_NEUTRINO 1
 #define READ_SDT_CHARSETS 1
-//#define CLOCKFIX 1
 #define IRDETO_GUESSING 1
-//#define CS_ANTICASC 1
+#define CS_ANTICASC 1
 #define WITH_DEBUG 1
 #define WITH_LB 1
-//#define CS_CACHEEX 1
+#define CS_CACHEEX 1
 //#define CS_CACHEEX_AIO 1
-//#define CW_CYCLE_CHECK 1
+#define CW_CYCLE_CHECK 1
 //#define LCDSUPPORT 1
 //#define LEDSUPPORT 1
-//#define IPV6SUPPORT 1
+#define IPV6SUPPORT 1
 #define MODULE_MONITOR 1
+#define WITH_CARDLIST 1
 
 //#define MODULE_CAMD33 1
 #define MODULE_CAMD35 1
@@ -35,10 +35,10 @@
 #define MODULE_CCCSHARE 1
 #define MODULE_GBOX 1
 #define MODULE_RADEGAST 1
-//#define MODULE_SERIAL 1
-//#define MODULE_CONSTCW 1
-//#define MODULE_PANDORA 1
-//#define MODULE_GHTTP 1
+#define MODULE_SERIAL 1
+#define MODULE_CONSTCW 1
+#define MODULE_PANDORA 1
+#define MODULE_GHTTP 1
 #define MODULE_SCAM 1
 
 #define WITH_CARDREADER 1
@@ -61,12 +61,12 @@
 
 #define CARDREADER_PHOENIX 1
 #define CARDREADER_INTERNAL 1
-//#define CARDREADER_MP35 1
-//#define CARDREADER_SC8IN1 1
-//#define CARDREADER_SMARGO 1
-//#define CARDREADER_DB2COM 1
+#define CARDREADER_MP35 1
+#define CARDREADER_SC8IN1 1
+#define CARDREADER_SMARGO 1
+#define CARDREADER_DB2COM 1
 #define CARDREADER_STINGER 1
-//#define CARDREADER_DRECAS 1
+#define CARDREADER_DRECAS 1
 
 #ifdef WITH_PCSC
 #define CARDREADER_PCSC 1
@@ -89,11 +89,11 @@
 #endif
 
 #ifdef WITH_STAPI
-//#define CARDREADER_STAPI 1
+#define CARDREADER_STAPI 1
 #endif
 
 #ifdef WITH_STAPI5
-//#define CARDREADER_STAPI5 1
+#define CARDREADER_STAPI5 1
 #endif
 
 #ifdef READER_DRE
diff -Nur oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/config.sh oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/config.sh
--- oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/config.sh	2022-12-31 12:30:12.008777882 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/config.sh	2022-12-31 13:46:06.471044138 +0100
@@ -1,6 +1,6 @@
 #!/bin/sh
 
-addons="WEBIF WEBIF_LIVELOG WEBIF_JQUERY TOUCH WITH_SSL HAVE_DVBAPI WITH_NEUTRINO READ_SDT_CHARSETS IRDETO_GUESSING CS_ANTICASC WITH_DEBUG MODULE_MONITOR WITH_LB CS_CACHEEX CS_CACHEEX_AIO CW_CYCLE_CHECK LCDSUPPORT LEDSUPPORT CLOCKFIX IPV6SUPPORT WITH_EMU WITH_SOFTCAM WITH_ARM_NEON"
+addons="WEBIF WEBIF_LIVELOG WEBIF_JQUERY TOUCH WITH_SSL HAVE_DVBAPI WITH_NEUTRINO READ_SDT_CHARSETS IRDETO_GUESSING CS_ANTICASC WITH_DEBUG MODULE_MONITOR WITH_LB CS_CACHEEX CS_CACHEEX_AIO CW_CYCLE_CHECK LCDSUPPORT LEDSUPPORT IPV6SUPPORT WITH_CARDLIST WITH_EMU WITH_SOFTCAM WITH_ARM_NEON"
 protocols="MODULE_CAMD33 MODULE_CAMD35 MODULE_CAMD35_TCP MODULE_NEWCAMD MODULE_CCCAM MODULE_CCCSHARE MODULE_GBOX MODULE_RADEGAST MODULE_SCAM MODULE_SERIAL MODULE_CONSTCW MODULE_PANDORA MODULE_GHTTP"
 readers="READER_NAGRA READER_NAGRA_MERLIN READER_IRDETO READER_CONAX READER_CRYPTOWORKS READER_SECA READER_VIACCESS READER_VIDEOGUARD READER_DRE READER_TONGFANG READER_BULCRYPT READER_GRIFFIN READER_DGCRYPT"
 card_readers="CARDREADER_PHOENIX CARDREADER_INTERNAL CARDREADER_SC8IN1 CARDREADER_MP35 CARDREADER_SMARGO CARDREADER_DB2COM CARDREADER_STAPI CARDREADER_STAPI5 CARDREADER_STINGER CARDREADER_DRECAS"
@@ -10,7 +10,7 @@
 CONFIG_WEBIF_LIVELOG=y
 CONFIG_WEBIF_JQUERY=y
 # CONFIG_TOUCH=n
-# CONFIG_WITH_SSL=n
+CONFIG_WITH_SSL=y
 CONFIG_HAVE_DVBAPI=y
 # CONFIG_WITH_NEUTRINO=n
 CONFIG_READ_SDT_CHARSETS=y
@@ -24,11 +24,11 @@
 # CONFIG_CW_CYCLE_CHECK=n
 # CONFIG_LCDSUPPORT=n
 # CONFIG_LEDSUPPORT=n
-# CONFIG_CLOCKFIX=n
-# CONFIG_IPV6SUPPORT=n
+CONFIG_IPV6SUPPORT=y
 CONFIG_WITH_EMU=y
-CONFIG_WITH_SOFTCAM=y
+# CONFIG_WITH_SOFTCAM=n
 CONFIG_WITH_ARM_NEON=y
+CONFIG_WITH_CARDLIST=y
 # CONFIG_MODULE_CAMD33=n
 CONFIG_MODULE_CAMD35=y
 CONFIG_MODULE_CAMD35_TCP=y
@@ -306,6 +306,7 @@
 	enabled WITH_EMU && enable_opt READER_VIACCESS >/dev/null
 	enabled WITH_EMU && enable_opt MODULE_NEWCAMD >/dev/null
 	disabled WITH_EMU && disable_opt WITH_SOFTCAM >/dev/null
+	disabled WITH_CARDREADER && disable_opt WITH_CARDLIST >/dev/null
 }
 
 list_config() {
@@ -472,11 +473,11 @@
 		CW_CYCLE_CHECK		"CW Cycle Check"						$(check_test "CW_CYCLE_CHECK") \
 		LCDSUPPORT			"LCD support"							$(check_test "LCDSUPPORT") \
 		LEDSUPPORT			"LED support"							$(check_test "LEDSUPPORT") \
-		CLOCKFIX			"Clockfix (disable on old systems!)"	$(check_test "CLOCKFIX") \
 		IPV6SUPPORT			"IPv6 support (experimental)"			$(check_test "IPV6SUPPORT") \
 		WITH_EMU			"Emulator support"						$(check_test "WITH_EMU") \
 		WITH_SOFTCAM		"Built-in SoftCam.Key"					$(check_test "WITH_SOFTCAM") \
 		WITH_ARM_NEON		"ARM Neon Support"						$(check_test "WITH_ARM_NEON") \
+		WITH_CARDLIST		"Cardlist support"						$(check_test "WITH_CARDLIST") \
 		2> ${tempfile}
 
 	opt=${?}
@@ -710,18 +711,18 @@
 		break
 	;;
 	'-r'|'--oscam-revision')
-		revision=`(svnversion -n . 2>/dev/null || printf 0) | sed 's/.*://; s/[^0-9]*$//; s/^$/0/'`
-		if [ "$revision" = "0" ]
-		then
-			which git > /dev/null 2>&1 && revision=`git log -25 --pretty=%B | grep git-svn-id | head -n 1 | sed -n -e 's/^.*trunk@\([0-9]*\) .*$/\1/p'`
-		fi
-		emuversion=`grep EMU_VERSION module-emulator-osemu.h | awk '{ print $3 }'`
-		if [ -z "$emuversion" ]
-		then
-			echo "$revision"
+		revision=$(echo $(git rev-list --count HEAD 2>/dev/null; git rev-list --all --max-count=1 2>/dev/null | cut -c1-7) | sed -e "s| |+|g")
+		if [ -n "$revision" ]; then
+			revision="git+$revision+svn$(cat .trunk-svn)+emu$(grep EMU_VERSION module-emulator-osemu.h | awk '{ print $3 }')"
 		else
-			echo "$revision-$emuversion"
+			revision=$(svn info 2>/dev/null | grep Revision | cut -d ' ' -f 2)
+			if [ -n "$revision" ]; then
+				revision="git+$revision+UNKNOWN+svn$(cat .trunk-svn)+emu$(grep EMU_VERSION module-emulator-osemu.h | awk '{ print $3 }')"
+			else
+				revision="git+UNKNOWN+UNKNOWN+svn$(cat .trunk-svn)+emu$(grep EMU_VERSION module-emulator-osemu.h | awk '{ print $3 }')"
+			fi
 		fi
+		echo $revision
 		break
 	;;
 	'--emu-version')
diff -Nur oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/csctapi/CMakeLists.txt oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/csctapi/CMakeLists.txt
--- oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/csctapi/CMakeLists.txt	2022-12-30 18:34:37.848461239 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/csctapi/CMakeLists.txt	2022-12-31 13:44:02.250047948 +0100
@@ -20,12 +20,12 @@
 endif(OSCamOperatingSystem MATCHES "Mac OS X" AND NOT HAVE_LIBUSB AND HAVE_PCSC)
 if (OSCamOperatingSystem MATCHES "Mac OS X")
 file (GLOB csctapi_mac "ifd_azbox*" "ifd_cool*" "ifd_stapi*" ${mac_usbpcsc})
-file (GLOB csctapi_srcs "*.c")
+file (GLOB csctapi_srcs "*.c" "../cardlist.c")
 list(REMOVE_ITEM csctapi_srcs ${csctapi_mac})
 file (GLOB csctapi_hdrs "*.h")
 list(REMOVE_ITEM csctapi_hdrs ${csctapi_mac})
 else (OSCamOperatingSystem MATCHES "Mac OS X")
-file (GLOB csctapi_srcs "*.c")
+file (GLOB csctapi_srcs "*.c" "../cardlist.c")
 file (GLOB csctapi_hdrs "*.h")
 endif (OSCamOperatingSystem MATCHES "Mac OS X")
 
diff -Nur oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/csctapi/icc_async.c oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/csctapi/icc_async.c
--- oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/csctapi/icc_async.c	2022-12-31 12:28:54.466660923 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/csctapi/icc_async.c	2022-12-31 13:44:02.330046013 +0100
@@ -7,13 +7,15 @@
 #include "io_serial.h"
 #include "ifd_phoenix.h"
 #include "../oscam-time.h"
-#include "cardlist.h"
 #ifdef READER_NAGRA_MERLIN
 #include "../cscrypt/fast_aes.h"
 #include "../cscrypt/sha256.h"
 #include "../cscrypt/mdc2.h"
 #include "../cscrypt/idea.h"
 #endif
+#ifdef WITH_CARDLIST
+#include "../cardlist.h"
+#endif
 
 #define OK 0
 #define ERROR 1
@@ -299,16 +301,16 @@
 	uint32_t atr_size;
 	ATR_GetRaw(atr, atrarr, &atr_size);
 	char tmp[atr_size * 3 + 1];
-	memcpy(current.atr, cs_hexdump(1, atrarr, atr_size, tmp, sizeof(tmp)), atr_size * 3 - 1);
-	current.atr[atr_size * 3 - 1] = '\0';
-	rdr_log(reader, "ATR: %s", current.atr);
+	rdr_log(reader, "ATR: %s", cs_hexdump(1, atrarr, atr_size, tmp, sizeof(tmp)));
 	memcpy(reader->card_atr, atrarr, atr_size);
 	reader->card_atr_length = atr_size;
+#ifdef WITH_CARDLIST
+	memcpy(current.atr, cs_hexdump(1, atrarr, atr_size, tmp, sizeof(tmp)), atr_size * 3 - 1);
 	findatr(reader);
 	if ( current.found == 1 ) {
-		rdr_log(reader, "%s recognized", current.providername);
+		rdr_log(reader, "%s %s", current.providername, current.info);
 	}
-
+#endif
 	// Get ICC reader->convention
 	if(ATR_GetConvention(atr, &(reader->convention)) != ATR_OK)
 	{
diff -Nur oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/globals.h oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/globals.h
--- oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/globals.h	2022-12-31 12:28:54.470660825 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/globals.h	2022-12-31 13:48:08.800085989 +0100
@@ -143,7 +143,7 @@
  * =========================== */
 // Prevent use of unsafe functions (doesn't work for MacOSX)
 #if !defined(__APPLE__)
-#define strcpy(a,b) UNSAFE_STRCPY_USE_CS_STRNCPY_INSTEAD()
+// #define strcpy(a,b) UNSAFE_STRCPY_USE_CS_STRNCPY_INSTEAD()
 #define sprintf(a,...) UNSAFE_SPRINTF_USE_SNPRINTF_INSTEAD()
 #define strtok(a,b,c) UNSAFE_STRTOK_USE_STRTOK_R_INSTEAD()
 #define gmtime(a) UNSAFE_GMTIME_NOT_THREADSAFE_USE_CS_GMTIME_R()
@@ -360,7 +360,7 @@
 /* ===========================
  *			constants
  * =========================== */
-#define CS_VERSION				"1.20_svn"
+#define CS_VERSION				"smod"
 #ifndef CS_SVN_VERSION
 # define CS_SVN_VERSION			"test"
 #endif
@@ -369,11 +369,17 @@
 #define CS_AIO_VERSION			CS_SVN_VERSION
 #endif
 #endif
+#ifndef CS_SMOD_VERSION
+#   define CS_SMOD_VERSION			"test"
+#endif
+#ifndef CS_SMOD_VERSION_HASH
+#   define CS_SMOD_VERSION_HASH		""
+#endif
 #ifndef CS_TARGET
-# define CS_TARGET				"unknown"
+# define CS_TARGET				"smod"
 #endif
 #ifndef CS_CONFDIR
-#define CS_CONFDIR				"/usr/local/etc"
+#define CS_CONFDIR				"/etc/tuxbox/config"
 #endif
 #ifndef CS_LOGFILE
 #define CS_LOGFILE				"/var/log/oscam.log"
diff -Nur oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/LICENSE oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/LICENSE
--- oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/LICENSE	1970-01-01 01:00:00.000000000 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/LICENSE	2022-12-31 13:44:02.250047948 +0100
@@ -0,0 +1,674 @@
+                    GNU GENERAL PUBLIC LICENSE
+                       Version 3, 29 June 2007
+
+ Copyright (C) 2007 Free Software Foundation, Inc. <http://fsf.org/>
+ Everyone is permitted to copy and distribute verbatim copies
+ of this license document, but changing it is not allowed.
+
+                            Preamble
+
+  The GNU General Public License is a free, copyleft license for
+software and other kinds of works.
+
+  The licenses for most software and other practical works are designed
+to take away your freedom to share and change the works.  By contrast,
+the GNU General Public License is intended to guarantee your freedom to
+share and change all versions of a program--to make sure it remains free
+software for all its users.  We, the Free Software Foundation, use the
+GNU General Public License for most of our software; it applies also to
+any other work released this way by its authors.  You can apply it to
+your programs, too.
+
+  When we speak of free software, we are referring to freedom, not
+price.  Our General Public Licenses are designed to make sure that you
+have the freedom to distribute copies of free software (and charge for
+them if you wish), that you receive source code or can get it if you
+want it, that you can change the software or use pieces of it in new
+free programs, and that you know you can do these things.
+
+  To protect your rights, we need to prevent others from denying you
+these rights or asking you to surrender the rights.  Therefore, you have
+certain responsibilities if you distribute copies of the software, or if
+you modify it: responsibilities to respect the freedom of others.
+
+  For example, if you distribute copies of such a program, whether
+gratis or for a fee, you must pass on to the recipients the same
+freedoms that you received.  You must make sure that they, too, receive
+or can get the source code.  And you must show them these terms so they
+know their rights.
+
+  Developers that use the GNU GPL protect your rights with two steps:
+(1) assert copyright on the software, and (2) offer you this License
+giving you legal permission to copy, distribute and/or modify it.
+
+  For the developers' and authors' protection, the GPL clearly explains
+that there is no warranty for this free software.  For both users' and
+authors' sake, the GPL requires that modified versions be marked as
+changed, so that their problems will not be attributed erroneously to
+authors of previous versions.
+
+  Some devices are designed to deny users access to install or run
+modified versions of the software inside them, although the manufacturer
+can do so.  This is fundamentally incompatible with the aim of
+protecting users' freedom to change the software.  The systematic
+pattern of such abuse occurs in the area of products for individuals to
+use, which is precisely where it is most unacceptable.  Therefore, we
+have designed this version of the GPL to prohibit the practice for those
+products.  If such problems arise substantially in other domains, we
+stand ready to extend this provision to those domains in future versions
+of the GPL, as needed to protect the freedom of users.
+
+  Finally, every program is threatened constantly by software patents.
+States should not allow patents to restrict development and use of
+software on general-purpose computers, but in those that do, we wish to
+avoid the special danger that patents applied to a free program could
+make it effectively proprietary.  To prevent this, the GPL assures that
+patents cannot be used to render the program non-free.
+
+  The precise terms and conditions for copying, distribution and
+modification follow.
+
+                       TERMS AND CONDITIONS
+
+  0. Definitions.
+
+  "This License" refers to version 3 of the GNU General Public License.
+
+  "Copyright" also means copyright-like laws that apply to other kinds of
+works, such as semiconductor masks.
+
+  "The Program" refers to any copyrightable work licensed under this
+License.  Each licensee is addressed as "you".  "Licensees" and
+"recipients" may be individuals or organizations.
+
+  To "modify" a work means to copy from or adapt all or part of the work
+in a fashion requiring copyright permission, other than the making of an
+exact copy.  The resulting work is called a "modified version" of the
+earlier work or a work "based on" the earlier work.
+
+  A "covered work" means either the unmodified Program or a work based
+on the Program.
+
+  To "propagate" a work means to do anything with it that, without
+permission, would make you directly or secondarily liable for
+infringement under applicable copyright law, except executing it on a
+computer or modifying a private copy.  Propagation includes copying,
+distribution (with or without modification), making available to the
+public, and in some countries other activities as well.
+
+  To "convey" a work means any kind of propagation that enables other
+parties to make or receive copies.  Mere interaction with a user through
+a computer network, with no transfer of a copy, is not conveying.
+
+  An interactive user interface displays "Appropriate Legal Notices"
+to the extent that it includes a convenient and prominently visible
+feature that (1) displays an appropriate copyright notice, and (2)
+tells the user that there is no warranty for the work (except to the
+extent that warranties are provided), that licensees may convey the
+work under this License, and how to view a copy of this License.  If
+the interface presents a list of user commands or options, such as a
+menu, a prominent item in the list meets this criterion.
+
+  1. Source Code.
+
+  The "source code" for a work means the preferred form of the work
+for making modifications to it.  "Object code" means any non-source
+form of a work.
+
+  A "Standard Interface" means an interface that either is an official
+standard defined by a recognized standards body, or, in the case of
+interfaces specified for a particular programming language, one that
+is widely used among developers working in that language.
+
+  The "System Libraries" of an executable work include anything, other
+than the work as a whole, that (a) is included in the normal form of
+packaging a Major Component, but which is not part of that Major
+Component, and (b) serves only to enable use of the work with that
+Major Component, or to implement a Standard Interface for which an
+implementation is available to the public in source code form.  A
+"Major Component", in this context, means a major essential component
+(kernel, window system, and so on) of the specific operating system
+(if any) on which the executable work runs, or a compiler used to
+produce the work, or an object code interpreter used to run it.
+
+  The "Corresponding Source" for a work in object code form means all
+the source code needed to generate, install, and (for an executable
+work) run the object code and to modify the work, including scripts to
+control those activities.  However, it does not include the work's
+System Libraries, or general-purpose tools or generally available free
+programs which are used unmodified in performing those activities but
+which are not part of the work.  For example, Corresponding Source
+includes interface definition files associated with source files for
+the work, and the source code for shared libraries and dynamically
+linked subprograms that the work is specifically designed to require,
+such as by intimate data communication or control flow between those
+subprograms and other parts of the work.
+
+  The Corresponding Source need not include anything that users
+can regenerate automatically from other parts of the Corresponding
+Source.
+
+  The Corresponding Source for a work in source code form is that
+same work.
+
+  2. Basic Permissions.
+
+  All rights granted under this License are granted for the term of
+copyright on the Program, and are irrevocable provided the stated
+conditions are met.  This License explicitly affirms your unlimited
+permission to run the unmodified Program.  The output from running a
+covered work is covered by this License only if the output, given its
+content, constitutes a covered work.  This License acknowledges your
+rights of fair use or other equivalent, as provided by copyright law.
+
+  You may make, run and propagate covered works that you do not
+convey, without conditions so long as your license otherwise remains
+in force.  You may convey covered works to others for the sole purpose
+of having them make modifications exclusively for you, or provide you
+with facilities for running those works, provided that you comply with
+the terms of this License in conveying all material for which you do
+not control copyright.  Those thus making or running the covered works
+for you must do so exclusively on your behalf, under your direction
+and control, on terms that prohibit them from making any copies of
+your copyrighted material outside their relationship with you.
+
+  Conveying under any other circumstances is permitted solely under
+the conditions stated below.  Sublicensing is not allowed; section 10
+makes it unnecessary.
+
+  3. Protecting Users' Legal Rights From Anti-Circumvention Law.
+
+  No covered work shall be deemed part of an effective technological
+measure under any applicable law fulfilling obligations under article
+11 of the WIPO copyright treaty adopted on 20 December 1996, or
+similar laws prohibiting or restricting circumvention of such
+measures.
+
+  When you convey a covered work, you waive any legal power to forbid
+circumvention of technological measures to the extent such circumvention
+is effected by exercising rights under this License with respect to
+the covered work, and you disclaim any intention to limit operation or
+modification of the work as a means of enforcing, against the work's
+users, your or third parties' legal rights to forbid circumvention of
+technological measures.
+
+  4. Conveying Verbatim Copies.
+
+  You may convey verbatim copies of the Program's source code as you
+receive it, in any medium, provided that you conspicuously and
+appropriately publish on each copy an appropriate copyright notice;
+keep intact all notices stating that this License and any
+non-permissive terms added in accord with section 7 apply to the code;
+keep intact all notices of the absence of any warranty; and give all
+recipients a copy of this License along with the Program.
+
+  You may charge any price or no price for each copy that you convey,
+and you may offer support or warranty protection for a fee.
+
+  5. Conveying Modified Source Versions.
+
+  You may convey a work based on the Program, or the modifications to
+produce it from the Program, in the form of source code under the
+terms of section 4, provided that you also meet all of these conditions:
+
+    a) The work must carry prominent notices stating that you modified
+    it, and giving a relevant date.
+
+    b) The work must carry prominent notices stating that it is
+    released under this License and any conditions added under section
+    7.  This requirement modifies the requirement in section 4 to
+    "keep intact all notices".
+
+    c) You must license the entire work, as a whole, under this
+    License to anyone who comes into possession of a copy.  This
+    License will therefore apply, along with any applicable section 7
+    additional terms, to the whole of the work, and all its parts,
+    regardless of how they are packaged.  This License gives no
+    permission to license the work in any other way, but it does not
+    invalidate such permission if you have separately received it.
+
+    d) If the work has interactive user interfaces, each must display
+    Appropriate Legal Notices; however, if the Program has interactive
+    interfaces that do not display Appropriate Legal Notices, your
+    work need not make them do so.
+
+  A compilation of a covered work with other separate and independent
+works, which are not by their nature extensions of the covered work,
+and which are not combined with it such as to form a larger program,
+in or on a volume of a storage or distribution medium, is called an
+"aggregate" if the compilation and its resulting copyright are not
+used to limit the access or legal rights of the compilation's users
+beyond what the individual works permit.  Inclusion of a covered work
+in an aggregate does not cause this License to apply to the other
+parts of the aggregate.
+
+  6. Conveying Non-Source Forms.
+
+  You may convey a covered work in object code form under the terms
+of sections 4 and 5, provided that you also convey the
+machine-readable Corresponding Source under the terms of this License,
+in one of these ways:
+
+    a) Convey the object code in, or embodied in, a physical product
+    (including a physical distribution medium), accompanied by the
+    Corresponding Source fixed on a durable physical medium
+    customarily used for software interchange.
+
+    b) Convey the object code in, or embodied in, a physical product
+    (including a physical distribution medium), accompanied by a
+    written offer, valid for at least three years and valid for as
+    long as you offer spare parts or customer support for that product
+    model, to give anyone who possesses the object code either (1) a
+    copy of the Corresponding Source for all the software in the
+    product that is covered by this License, on a durable physical
+    medium customarily used for software interchange, for a price no
+    more than your reasonable cost of physically performing this
+    conveying of source, or (2) access to copy the
+    Corresponding Source from a network server at no charge.
+
+    c) Convey individual copies of the object code with a copy of the
+    written offer to provide the Corresponding Source.  This
+    alternative is allowed only occasionally and noncommercially, and
+    only if you received the object code with such an offer, in accord
+    with subsection 6b.
+
+    d) Convey the object code by offering access from a designated
+    place (gratis or for a charge), and offer equivalent access to the
+    Corresponding Source in the same way through the same place at no
+    further charge.  You need not require recipients to copy the
+    Corresponding Source along with the object code.  If the place to
+    copy the object code is a network server, the Corresponding Source
+    may be on a different server (operated by you or a third party)
+    that supports equivalent copying facilities, provided you maintain
+    clear directions next to the object code saying where to find the
+    Corresponding Source.  Regardless of what server hosts the
+    Corresponding Source, you remain obligated to ensure that it is
+    available for as long as needed to satisfy these requirements.
+
+    e) Convey the object code using peer-to-peer transmission, provided
+    you inform other peers where the object code and Corresponding
+    Source of the work are being offered to the general public at no
+    charge under subsection 6d.
+
+  A separable portion of the object code, whose source code is excluded
+from the Corresponding Source as a System Library, need not be
+included in conveying the object code work.
+
+  A "User Product" is either (1) a "consumer product", which means any
+tangible personal property which is normally used for personal, family,
+or household purposes, or (2) anything designed or sold for incorporation
+into a dwelling.  In determining whether a product is a consumer product,
+doubtful cases shall be resolved in favor of coverage.  For a particular
+product received by a particular user, "normally used" refers to a
+typical or common use of that class of product, regardless of the status
+of the particular user or of the way in which the particular user
+actually uses, or expects or is expected to use, the product.  A product
+is a consumer product regardless of whether the product has substantial
+commercial, industrial or non-consumer uses, unless such uses represent
+the only significant mode of use of the product.
+
+  "Installation Information" for a User Product means any methods,
+procedures, authorization keys, or other information required to install
+and execute modified versions of a covered work in that User Product from
+a modified version of its Corresponding Source.  The information must
+suffice to ensure that the continued functioning of the modified object
+code is in no case prevented or interfered with solely because
+modification has been made.
+
+  If you convey an object code work under this section in, or with, or
+specifically for use in, a User Product, and the conveying occurs as
+part of a transaction in which the right of possession and use of the
+User Product is transferred to the recipient in perpetuity or for a
+fixed term (regardless of how the transaction is characterized), the
+Corresponding Source conveyed under this section must be accompanied
+by the Installation Information.  But this requirement does not apply
+if neither you nor any third party retains the ability to install
+modified object code on the User Product (for example, the work has
+been installed in ROM).
+
+  The requirement to provide Installation Information does not include a
+requirement to continue to provide support service, warranty, or updates
+for a work that has been modified or installed by the recipient, or for
+the User Product in which it has been modified or installed.  Access to a
+network may be denied when the modification itself materially and
+adversely affects the operation of the network or violates the rules and
+protocols for communication across the network.
+
+  Corresponding Source conveyed, and Installation Information provided,
+in accord with this section must be in a format that is publicly
+documented (and with an implementation available to the public in
+source code form), and must require no special password or key for
+unpacking, reading or copying.
+
+  7. Additional Terms.
+
+  "Additional permissions" are terms that supplement the terms of this
+License by making exceptions from one or more of its conditions.
+Additional permissions that are applicable to the entire Program shall
+be treated as though they were included in this License, to the extent
+that they are valid under applicable law.  If additional permissions
+apply only to part of the Program, that part may be used separately
+under those permissions, but the entire Program remains governed by
+this License without regard to the additional permissions.
+
+  When you convey a copy of a covered work, you may at your option
+remove any additional permissions from that copy, or from any part of
+it.  (Additional permissions may be written to require their own
+removal in certain cases when you modify the work.)  You may place
+additional permissions on material, added by you to a covered work,
+for which you have or can give appropriate copyright permission.
+
+  Notwithstanding any other provision of this License, for material you
+add to a covered work, you may (if authorized by the copyright holders of
+that material) supplement the terms of this License with terms:
+
+    a) Disclaiming warranty or limiting liability differently from the
+    terms of sections 15 and 16 of this License; or
+
+    b) Requiring preservation of specified reasonable legal notices or
+    author attributions in that material or in the Appropriate Legal
+    Notices displayed by works containing it; or
+
+    c) Prohibiting misrepresentation of the origin of that material, or
+    requiring that modified versions of such material be marked in
+    reasonable ways as different from the original version; or
+
+    d) Limiting the use for publicity purposes of names of licensors or
+    authors of the material; or
+
+    e) Declining to grant rights under trademark law for use of some
+    trade names, trademarks, or service marks; or
+
+    f) Requiring indemnification of licensors and authors of that
+    material by anyone who conveys the material (or modified versions of
+    it) with contractual assumptions of liability to the recipient, for
+    any liability that these contractual assumptions directly impose on
+    those licensors and authors.
+
+  All other non-permissive additional terms are considered "further
+restrictions" within the meaning of section 10.  If the Program as you
+received it, or any part of it, contains a notice stating that it is
+governed by this License along with a term that is a further
+restriction, you may remove that term.  If a license document contains
+a further restriction but permits relicensing or conveying under this
+License, you may add to a covered work material governed by the terms
+of that license document, provided that the further restriction does
+not survive such relicensing or conveying.
+
+  If you add terms to a covered work in accord with this section, you
+must place, in the relevant source files, a statement of the
+additional terms that apply to those files, or a notice indicating
+where to find the applicable terms.
+
+  Additional terms, permissive or non-permissive, may be stated in the
+form of a separately written license, or stated as exceptions;
+the above requirements apply either way.
+
+  8. Termination.
+
+  You may not propagate or modify a covered work except as expressly
+provided under this License.  Any attempt otherwise to propagate or
+modify it is void, and will automatically terminate your rights under
+this License (including any patent licenses granted under the third
+paragraph of section 11).
+
+  However, if you cease all violation of this License, then your
+license from a particular copyright holder is reinstated (a)
+provisionally, unless and until the copyright holder explicitly and
+finally terminates your license, and (b) permanently, if the copyright
+holder fails to notify you of the violation by some reasonable means
+prior to 60 days after the cessation.
+
+  Moreover, your license from a particular copyright holder is
+reinstated permanently if the copyright holder notifies you of the
+violation by some reasonable means, this is the first time you have
+received notice of violation of this License (for any work) from that
+copyright holder, and you cure the violation prior to 30 days after
+your receipt of the notice.
+
+  Termination of your rights under this section does not terminate the
+licenses of parties who have received copies or rights from you under
+this License.  If your rights have been terminated and not permanently
+reinstated, you do not qualify to receive new licenses for the same
+material under section 10.
+
+  9. Acceptance Not Required for Having Copies.
+
+  You are not required to accept this License in order to receive or
+run a copy of the Program.  Ancillary propagation of a covered work
+occurring solely as a consequence of using peer-to-peer transmission
+to receive a copy likewise does not require acceptance.  However,
+nothing other than this License grants you permission to propagate or
+modify any covered work.  These actions infringe copyright if you do
+not accept this License.  Therefore, by modifying or propagating a
+covered work, you indicate your acceptance of this License to do so.
+
+  10. Automatic Licensing of Downstream Recipients.
+
+  Each time you convey a covered work, the recipient automatically
+receives a license from the original licensors, to run, modify and
+propagate that work, subject to this License.  You are not responsible
+for enforcing compliance by third parties with this License.
+
+  An "entity transaction" is a transaction transferring control of an
+organization, or substantially all assets of one, or subdividing an
+organization, or merging organizations.  If propagation of a covered
+work results from an entity transaction, each party to that
+transaction who receives a copy of the work also receives whatever
+licenses to the work the party's predecessor in interest had or could
+give under the previous paragraph, plus a right to possession of the
+Corresponding Source of the work from the predecessor in interest, if
+the predecessor has it or can get it with reasonable efforts.
+
+  You may not impose any further restrictions on the exercise of the
+rights granted or affirmed under this License.  For example, you may
+not impose a license fee, royalty, or other charge for exercise of
+rights granted under this License, and you may not initiate litigation
+(including a cross-claim or counterclaim in a lawsuit) alleging that
+any patent claim is infringed by making, using, selling, offering for
+sale, or importing the Program or any portion of it.
+
+  11. Patents.
+
+  A "contributor" is a copyright holder who authorizes use under this
+License of the Program or a work on which the Program is based.  The
+work thus licensed is called the contributor's "contributor version".
+
+  A contributor's "essential patent claims" are all patent claims
+owned or controlled by the contributor, whether already acquired or
+hereafter acquired, that would be infringed by some manner, permitted
+by this License, of making, using, or selling its contributor version,
+but do not include claims that would be infringed only as a
+consequence of further modification of the contributor version.  For
+purposes of this definition, "control" includes the right to grant
+patent sublicenses in a manner consistent with the requirements of
+this License.
+
+  Each contributor grants you a non-exclusive, worldwide, royalty-free
+patent license under the contributor's essential patent claims, to
+make, use, sell, offer for sale, import and otherwise run, modify and
+propagate the contents of its contributor version.
+
+  In the following three paragraphs, a "patent license" is any express
+agreement or commitment, however denominated, not to enforce a patent
+(such as an express permission to practice a patent or covenant not to
+sue for patent infringement).  To "grant" such a patent license to a
+party means to make such an agreement or commitment not to enforce a
+patent against the party.
+
+  If you convey a covered work, knowingly relying on a patent license,
+and the Corresponding Source of the work is not available for anyone
+to copy, free of charge and under the terms of this License, through a
+publicly available network server or other readily accessible means,
+then you must either (1) cause the Corresponding Source to be so
+available, or (2) arrange to deprive yourself of the benefit of the
+patent license for this particular work, or (3) arrange, in a manner
+consistent with the requirements of this License, to extend the patent
+license to downstream recipients.  "Knowingly relying" means you have
+actual knowledge that, but for the patent license, your conveying the
+covered work in a country, or your recipient's use of the covered work
+in a country, would infringe one or more identifiable patents in that
+country that you have reason to believe are valid.
+
+  If, pursuant to or in connection with a single transaction or
+arrangement, you convey, or propagate by procuring conveyance of, a
+covered work, and grant a patent license to some of the parties
+receiving the covered work authorizing them to use, propagate, modify
+or convey a specific copy of the covered work, then the patent license
+you grant is automatically extended to all recipients of the covered
+work and works based on it.
+
+  A patent license is "discriminatory" if it does not include within
+the scope of its coverage, prohibits the exercise of, or is
+conditioned on the non-exercise of one or more of the rights that are
+specifically granted under this License.  You may not convey a covered
+work if you are a party to an arrangement with a third party that is
+in the business of distributing software, under which you make payment
+to the third party based on the extent of your activity of conveying
+the work, and under which the third party grants, to any of the
+parties who would receive the covered work from you, a discriminatory
+patent license (a) in connection with copies of the covered work
+conveyed by you (or copies made from those copies), or (b) primarily
+for and in connection with specific products or compilations that
+contain the covered work, unless you entered into that arrangement,
+or that patent license was granted, prior to 28 March 2007.
+
+  Nothing in this License shall be construed as excluding or limiting
+any implied license or other defenses to infringement that may
+otherwise be available to you under applicable patent law.
+
+  12. No Surrender of Others' Freedom.
+
+  If conditions are imposed on you (whether by court order, agreement or
+otherwise) that contradict the conditions of this License, they do not
+excuse you from the conditions of this License.  If you cannot convey a
+covered work so as to satisfy simultaneously your obligations under this
+License and any other pertinent obligations, then as a consequence you may
+not convey it at all.  For example, if you agree to terms that obligate you
+to collect a royalty for further conveying from those to whom you convey
+the Program, the only way you could satisfy both those terms and this
+License would be to refrain entirely from conveying the Program.
+
+  13. Use with the GNU Affero General Public License.
+
+  Notwithstanding any other provision of this License, you have
+permission to link or combine any covered work with a work licensed
+under version 3 of the GNU Affero General Public License into a single
+combined work, and to convey the resulting work.  The terms of this
+License will continue to apply to the part which is the covered work,
+but the special requirements of the GNU Affero General Public License,
+section 13, concerning interaction through a network will apply to the
+combination as such.
+
+  14. Revised Versions of this License.
+
+  The Free Software Foundation may publish revised and/or new versions of
+the GNU General Public License from time to time.  Such new versions will
+be similar in spirit to the present version, but may differ in detail to
+address new problems or concerns.
+
+  Each version is given a distinguishing version number.  If the
+Program specifies that a certain numbered version of the GNU General
+Public License "or any later version" applies to it, you have the
+option of following the terms and conditions either of that numbered
+version or of any later version published by the Free Software
+Foundation.  If the Program does not specify a version number of the
+GNU General Public License, you may choose any version ever published
+by the Free Software Foundation.
+
+  If the Program specifies that a proxy can decide which future
+versions of the GNU General Public License can be used, that proxy's
+public statement of acceptance of a version permanently authorizes you
+to choose that version for the Program.
+
+  Later license versions may give you additional or different
+permissions.  However, no additional obligations are imposed on any
+author or copyright holder as a result of your choosing to follow a
+later version.
+
+  15. Disclaimer of Warranty.
+
+  THERE IS NO WARRANTY FOR THE PROGRAM, TO THE EXTENT PERMITTED BY
+APPLICABLE LAW.  EXCEPT WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT
+HOLDERS AND/OR OTHER PARTIES PROVIDE THE PROGRAM "AS IS" WITHOUT WARRANTY
+OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO,
+THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+PURPOSE.  THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE PROGRAM
+IS WITH YOU.  SHOULD THE PROGRAM PROVE DEFECTIVE, YOU ASSUME THE COST OF
+ALL NECESSARY SERVICING, REPAIR OR CORRECTION.
+
+  16. Limitation of Liability.
+
+  IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
+WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MODIFIES AND/OR CONVEYS
+THE PROGRAM AS PERMITTED ABOVE, BE LIABLE TO YOU FOR DAMAGES, INCLUDING ANY
+GENERAL, SPECIAL, INCIDENTAL OR CONSEQUENTIAL DAMAGES ARISING OUT OF THE
+USE OR INABILITY TO USE THE PROGRAM (INCLUDING BUT NOT LIMITED TO LOSS OF
+DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD
+PARTIES OR A FAILURE OF THE PROGRAM TO OPERATE WITH ANY OTHER PROGRAMS),
+EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF
+SUCH DAMAGES.
+
+  17. Interpretation of Sections 15 and 16.
+
+  If the disclaimer of warranty and limitation of liability provided
+above cannot be given local legal effect according to their terms,
+reviewing courts shall apply local law that most closely approximates
+an absolute waiver of all civil liability in connection with the
+Program, unless a warranty or assumption of liability accompanies a
+copy of the Program in return for a fee.
+
+                     END OF TERMS AND CONDITIONS
+
+            How to Apply These Terms to Your New Programs
+
+  If you develop a new program, and you want it to be of the greatest
+possible use to the public, the best way to achieve this is to make it
+free software which everyone can redistribute and change under these terms.
+
+  To do so, attach the following notices to the program.  It is safest
+to attach them to the start of each source file to most effectively
+state the exclusion of warranty; and each file should have at least
+the "copyright" line and a pointer to where the full notice is found.
+
+    {one line to give the program's name and a brief idea of what it does.}
+    Copyright (C) {year}  {name of author}
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+Also add information on how to contact you by electronic and paper mail.
+
+  If the program does terminal interaction, make it output a short
+notice like this when it starts in an interactive mode:
+
+    {project}  Copyright (C) {year}  {fullname}
+    This program comes with ABSOLUTELY NO WARRANTY; for details type `show w'.
+    This is free software, and you are welcome to redistribute it
+    under certain conditions; type `show c' for details.
+
+The hypothetical commands `show w' and `show c' should show the appropriate
+parts of the General Public License.  Of course, your program's commands
+might be different; for a GUI interface, you would use an "about box".
+
+  You should also get your employer (if you work as a programmer) or school,
+if any, to sign a "copyright disclaimer" for the program, if necessary.
+For more information on this, and how to apply and follow the GNU GPL, see
+<http://www.gnu.org/licenses/>.
+
+  The GNU General Public License does not permit incorporating your program
+into proprietary programs.  If your program is a subroutine library, you
+may consider it more useful to permit linking proprietary applications with
+the library.  If this is what you want to do, use the GNU Lesser General
+Public License instead of this License.  But first, please read
+<http://www.gnu.org/philosophy/why-not-lgpl.html>.
diff -Nur oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/Makefile oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/Makefile
--- oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/Makefile	2022-12-30 19:23:46.313945320 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/Makefile	2022-12-31 13:44:02.330046013 +0100
@@ -4,8 +4,8 @@
 .SUFFIXES: .o .c
 .PHONY: all tests help README.build README.config simple default debug config menuconfig allyesconfig allnoconfig defconfig clean distclean
 
-VER     := $(shell ./config.sh --oscam-version)
-SVN_REV := $(shell ./config.sh --oscam-revision)
+VER      := $(shell ./config.sh --oscam-version)
+SMOD_REV := $(shell ./config.sh --oscam-revision | cut -d "+" -f1-3)
 
 uname_S := $(shell sh -c 'uname -s 2>/dev/null || echo not')
 
@@ -41,21 +41,18 @@
 LIB_DL = -ldl
 
 LIB_RT :=
-ifeq ($(uname_S),Linux)
-	ifeq "$(shell ./config.sh --enabled CLOCKFIX)" "Y"
-		LIB_RT := -lrt
-	endif
-endif
 ifeq ($(uname_S),FreeBSD)
 	LIB_DL :=
 endif
 
 override STD_LIBS := -lm $(LIB_PTHREAD) $(LIB_DL) $(LIB_RT)
-override STD_DEFS := -D'CS_SVN_VERSION="$(SVN_REV)"'
+override STD_DEFS := -D'CS_SVN_VERSION="$(shell ./config.sh --oscam-revision | cut -d "+" -f4 | sed -e "s/svn//")"'
+override STD_DEFS += -D'CS_SMOD_VERSION="$(shell ./config.sh --oscam-revision | cut -d "+" -f1-3)"'
+override STD_DEFS += -D'CS_SMOD_VERSION_HASH="$(shell ./config.sh --oscam-revision | cut -d "+" -f3)"'
 override STD_DEFS += -D'CS_CONFDIR="$(CONF_DIR)"'
 
 # Compiler warnings
-CC_WARN = -W -Wall -Wshadow -Wredundant-decls -Wstrict-prototypes -Wold-style-definition -Wno-unused-result
+CC_WARN = -W -Wall -Wshadow -Wredundant-decls -Wstrict-prototypes -Wold-style-definition
 
 # Compiler optimizations
 CC_OPTS = -O3 -ggdb -pipe -ffunction-sections -fdata-sections -funroll-loops -fomit-frame-pointer -fno-schedule-insns
@@ -63,7 +60,7 @@
 CC = $(CROSS_DIR)$(CROSS)gcc
 STRIP = $(CROSS_DIR)$(CROSS)strip
 
-LDFLAGS = -Wl,--gc-sections
+LDFLAGS += -Wl,--gc-sections
 
 TARGETHELP := $(shell $(CC) --target-help 2>&1)
 ifneq (,$(findstring sse2,$(TARGETHELP)))
@@ -74,6 +71,13 @@
 override CFLAGS += -fexpensive-optimizations
 endif
 
+TARGETHELP := $(shell $(CC) --target-help 2>&1)
+ifneq (,$(findstring sse2,$(TARGETHELP)))
+override CFLAGS += -fexpensive-optimizations -mmmx -msse -msse2 -msse3
+else
+override CFLAGS += -fexpensive-optimizations
+endif
+
 # The linker for powerpc have bug that prevents --gc-sections from working
 # Check for the linker version and if it matches disable --gc-sections
 # For more information about the bug see:
@@ -141,7 +145,7 @@
 endif
 
 ifeq ($(uname_S),Cygwin)
-	DEFAULT_PCSC_LIB += -lwinscard
+	DEFAULT_PCSC_LIB = -lwinscard
 endif
 
 DEFAULT_UTF8_FLAGS = -DWITH_UTF8
@@ -215,9 +219,9 @@
 # These variables will be used to select only needed files for compilation
 -include $(OBJDIR)/config.mak
 
-OSCAM_BIN := $(BINDIR)/oscam-$(VER)$(SVN_REV)-$(subst cygwin,cygwin.exe,$(TARGET))
+OSCAM_BIN := $(BINDIR)/oscam-$(VER)+$(SMOD_REV)-$(subst cygwin,cygwin.exe,$(TARGET))
 TESTS_BIN := tests.bin
-LIST_SMARGO_BIN := $(BINDIR)/list_smargo-$(VER)$(SVN_REV)-$(subst cygwin,cygwin.exe,$(TARGET))
+LIST_SMARGO_BIN := $(BINDIR)/list_smargo-$(VER)+$(SMOD_REV)-$(subst cygwin,cygwin.exe,$(TARGET))
 
 # Build list_smargo-.... only when WITH_LIBUSB build is requested.
 ifndef USE_LIBUSB
@@ -248,6 +252,7 @@
 SRC-$(CONFIG_LIB_FAST_AES) += cscrypt/fast_aes.c
 SRC-$(CONFIG_LIB_SHA256) += cscrypt/sha256.c
 
+SRC-$(CONFIG_WITH_CARDLIST) += cardlist.c
 SRC-$(CONFIG_WITH_CARDREADER) += csctapi/atr.c
 SRC-$(CONFIG_WITH_CARDREADER) += csctapi/icc_async.c
 SRC-$(CONFIG_WITH_CARDREADER) += csctapi/io_serial.c
@@ -409,7 +414,7 @@
 	@-mkdir -p $(OBJDIR)/cscrypt $(OBJDIR)/csctapi $(OBJDIR)/minilzo $(OBJDIR)/ffdecsa $(OBJDIR)/webif
 	@-printf "\
 +-------------------------------------------------------------------------------\n\
-| OSCam ver: $(VER) rev: $(SVN_REV) target: $(TARGET)\n\
+| OSCam ver: $(VER) rev: $(SMOD_REV) target: $(TARGET)\n\
 | Tools:\n\
 |  CROSS    = $(CROSS_DIR)$(CROSS)\n\
 |  CC       = $(CC)\n\
@@ -714,8 +719,8 @@
 \n\
    OSCAM_BIN=text  - This variable controls how the oscam binary will be named.\n\
                      Default OSCAM_BIN value is:\n\
-                      'BINDIR/oscam-VERSVN_REV-TARGET'\n\
-                     Once the variables (BINDIR, VER, SVN_REV and TARGET) are\n\
+                      'BINDIR/oscam-VER+SMOD_REV-TARGET'\n\
+                     Once the variables (BINDIR, VER, SMOD_REV and TARGET) are\n\
                      replaced, the resulting filename can look like this:\n\
                       'Distribution/oscam-1.20-unstable_svn7404-i486-slackware-linux-static'\n\
                      For example you can run: 'make OSCAM_BIN=my-oscam'\n\
diff -Nur oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/module-webif.c oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/module-webif.c
--- oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/module-webif.c	2022-12-31 12:28:54.502660049 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/module-webif.c	2022-12-31 14:31:17.981645161 +0100
@@ -6752,7 +6752,7 @@
 	tpl_addVar(vars, TPLADD, "DISPLAYREADERINFO", cfg.http_showreaderinfo ? "visible" : "hidden");
 	tpl_addVar(vars, TPLADD, "DISPLAYLOADINFO", cfg.http_showloadinfo ?"visible" : "hidden");
 	tpl_addVar(vars, TPLADD, "DISPLAYECMINFO", cfg.http_showecminfo ? "visible" : "hidden");
-	tpl_addVar(vars, TPLADD, "DISPLAYECMINFO_READERS", cfg.http_showecminfo ? "visible" : "hidden");
+	tpl_addVar(vars, TPLADD, "DISPLAYECMINFO_READERS", cfg.http_showreaderinfo ? "visible" : "hidden");
 
 	if(cfg.http_showcacheexinfo == 1 && config_enabled(CS_CACHEEX))
 	{
@@ -9320,6 +9320,8 @@
 
 			tpl_addVar(vars, TPLADD, "CS_VERSION", CS_VERSION);
 			tpl_addVar(vars, TPLADD, "CS_SVN_VERSION", CS_SVN_VERSION);
+			tpl_addVar(vars, TPLADD, "CS_SMOD_VERSION", CS_SMOD_VERSION);
+			tpl_addVar(vars, TPLADD, "CS_SMOD_VERSION_HASH", CS_SMOD_VERSION_HASH);
 			tpl_addVar(vars, TPLADD, "CS_TARGET", CS_TARGET);
 			tpl_addVar(vars, TPLADD, "HTTPOSCAMLABEL", xml_encode(vars,cfg.http_oscam_label));
 			if (!boxtype_is("generic"))
diff -Nur oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/module-webif-tpl.c oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/module-webif-tpl.c
--- oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/module-webif-tpl.c	2022-12-30 18:44:44.853956829 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/module-webif-tpl.c	2022-12-31 13:44:02.282047174 +0100
@@ -421,7 +421,6 @@
 											check_conf(HAVE_DVBAPI, ptr2);
 											check_conf(WITH_NEUTRINO, ptr2);
 											check_conf(READ_SDT_CHARSETS, ptr2);
-											check_conf(CLOCKFIX, ptr2);
 											check_conf(IPV6SUPPORT, ptr2);
 											check_conf(IRDETO_GUESSING, ptr2);
 											check_conf(LCDSUPPORT, ptr2);
diff -Nur oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/oscam.c oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/oscam.c
--- oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/oscam.c	2022-12-30 19:23:46.317945224 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/oscam.c	2022-12-31 13:44:02.282047174 +0100
@@ -341,7 +341,6 @@
 			if(config_enabled(WEBIF))
 			{
 				cs_http_use_utf8 = 1;
-				printf("WARNING: Web interface UTF-8 mode enabled. Carefully read documentation as bugs may arise.\n");
 			}
 			break;
 		case 'V': // --build-info
@@ -441,15 +440,10 @@
 	write_conf(CW_CYCLE_CHECK, "CW Cycle Check support");
 	write_conf(LCDSUPPORT, "LCD support");
 	write_conf(LEDSUPPORT, "LED support");
-	switch (cs_getclocktype())
-	{
-		case CLOCK_TYPE_UNKNOWN  : write_conf(CLOCKFIX, "Clockfix with UNKNOWN clock"); break;
-		case CLOCK_TYPE_REALTIME : write_conf(CLOCKFIX, "Clockfix with realtime clock"); break;
-		case CLOCK_TYPE_MONOTONIC: write_conf(CLOCKFIX, "Clockfix with monotonic clock"); break;
-	}
 	write_conf(IPV6SUPPORT, "IPv6 support");
 	write_conf(WITH_EMU, "Emulator support");
 	write_conf(WITH_SOFTCAM, "Built-in SoftCam.Key");
+	write_conf(WITH_CARDLIST, "Cardlist support");
 
 	fprintf(fp, "\n");
 	write_conf(MODULE_CAMD33, "camd 3.3x");
diff -Nur oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/oscam-chk.c oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/oscam-chk.c
--- oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/oscam-chk.c	2022-12-30 18:45:34.372766461 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/oscam-chk.c	2022-12-31 13:44:02.282047174 +0100
@@ -834,6 +834,26 @@
 
 uint8_t chk_if_ignore_checksum(ECM_REQUEST *er, FTAB *disablecrc_only_for)
 {
+	if(
+		(er->caid >> 8) == 0x0E
+		||
+		(er->caid == 0x0500 && ( er->prid == 0x030B00 || er->prid == 0x032830 || er->prid == 0x041950 || er->prid == 0x042800 || er->prid == 0x042820 || er->prid == 0x050F00 ))
+		||
+		(er->caid == 0x09C4 && er->prid == 0x000000)
+		||
+		(er->caid == 0x098C && er->prid == 0x000000)
+		||
+		(er->caid == 0x098D && er->prid == 0x000000)
+		||
+		(er->caid == 0x1811 && ( er->prid == 0x000000 || er->prid == 0x000007 || er->prid == 0x000107 || er->prid == 0x003311 || er->prid == 0x003315 || er->prid == 0x00331B || er->prid == 0x003341 || er->prid == 0x023311 ))
+		||
+		(er->caid == 0x1817 && ( er->prid == 0x000000 || er->prid == 0x00006A ))
+		||
+		(er->caid == 0x1818 && ( er->prid == 0x000000 || er->prid == 0x000007 || er->prid == 0x00006C ))
+		||
+		(er->caid == 0x1819 && ( er->prid == 0x000000 || er->prid == 0x000007 || er->prid == 0x00006D ))
+	) { return 1; }
+
 	if(!disablecrc_only_for->nfilts) { return 0; }
 
 	int32_t i, k;
diff -Nur oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/oscam-config-global.c oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/oscam-config-global.c
--- oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/oscam-config-global.c	2022-12-30 18:47:04.150608219 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/oscam-config-global.c	2022-12-31 13:44:02.278047271 +0100
@@ -16,7 +16,7 @@
 #define cs_conf "oscam.conf"
 
 #define DEFAULT_HTTP_PORT 8888
-#define DEFAULT_HTTP_ALLOW "127.0.0.1,192.168.0.0-192.168.255.255,10.0.0.0-10.255.255.255,172.16.0.0-172.31.255.255,::1"
+#define DEFAULT_HTTP_ALLOW "127.0.0.1,0.0.0.0-255.255.255.255,::1,fe80::-fe80::ffff:ffff:ffff:ffff,fd00::-fdff:ffff:ffff:ffff:ffff:ffff:ffff:ffff,2000::-3fff:ffff:ffff:ffff:ffff:ffff:ffff:ffff"
 
 static void disablelog_fn(const char *token, char *value, void *UNUSED(setting), FILE *f)
 {
@@ -341,24 +341,24 @@
 	DEF_OPT_INT32("bindwait"                       , OFS(bindwait)                      , CS_BIND_TIMEOUT),
 	DEF_OPT_UINT32("netprio"                       , OFS(netprio)                       , 0),
 	DEF_OPT_INT32("sleep"                          , OFS(tosleep)                       , 0),
-	DEF_OPT_INT32("unlockparental"                 , OFS(ulparent)                      , 0),
+	DEF_OPT_INT32("unlockparental"                 , OFS(ulparent)                      , 1),
 	DEF_OPT_INT32("nice"                           , OFS(nice)                          , 99),
 	DEF_OPT_INT32("maxlogsize"                     , OFS(max_log_size)                  , 10),
 	DEF_OPT_INT8("waitforcards"                    , OFS(waitforcards)                  , 1),
 	DEF_OPT_INT32("waitforcards_extra_delay"       , OFS(waitforcards_extra_delay)      , 500),
-	DEF_OPT_INT8("preferlocalcards"                , OFS(preferlocalcards)              , 0),
+	DEF_OPT_INT8("preferlocalcards"                , OFS(preferlocalcards)              , 1),
 	DEF_OPT_INT32("readerrestartseconds"           , OFS(reader_restart_seconds)        , 5),
 	DEF_OPT_INT8("dropdups"                        , OFS(dropdups)                      , 0),
-	DEF_OPT_INT8("reload_useraccounts"             , OFS(reload_useraccounts)           , 0),
-	DEF_OPT_INT8("reload_readers"                  , OFS(reload_readers)                , 0),
-	DEF_OPT_INT8("reload_provid"                   , OFS(reload_provid)                 , 0),
-	DEF_OPT_INT8("reload_services_ids"             , OFS(reload_services_ids)           , 0),
-	DEF_OPT_INT8("reload_tier_ids"                 , OFS(reload_tier_ids)               , 0),
-	DEF_OPT_INT8("reload_fakecws"                  , OFS(reload_fakecws)                , 0),
-	DEF_OPT_INT8("reload_ac_stat"                  , OFS(reload_ac_stat)                , 0),
-	DEF_OPT_INT8("reload_log"                      , OFS(reload_log)                    , 0),
-	DEF_OPT_INT8("block_same_ip"                   , OFS(block_same_ip)                 , 1),
-	DEF_OPT_INT8("block_same_name"                 , OFS(block_same_name)               , 1),
+	DEF_OPT_INT8("reload_useraccounts"             , OFS(reload_useraccounts)           , 1),
+	DEF_OPT_INT8("reload_readers"                  , OFS(reload_readers)                , 1),
+	DEF_OPT_INT8("reload_provid"                   , OFS(reload_provid)                 , 1),
+	DEF_OPT_INT8("reload_services_ids"             , OFS(reload_services_ids)           , 1),
+	DEF_OPT_INT8("reload_tier_ids"                 , OFS(reload_tier_ids)               , 1),
+	DEF_OPT_INT8("reload_fakecws"                  , OFS(reload_fakecws)                , 1),
+	DEF_OPT_INT8("reload_ac_stat"                  , OFS(reload_ac_stat)                , 1),
+	DEF_OPT_INT8("reload_log"                      , OFS(reload_log)                    , 1),
+	DEF_OPT_INT8("block_same_ip"                   , OFS(block_same_ip)                 , 0),
+	DEF_OPT_INT8("block_same_name"                 , OFS(block_same_name)               , 0),
 	DEF_OPT_STR("usrfile"                          , OFS(usrfile)                       , NULL),
 	DEF_OPT_STR("mailfile"                         , OFS(mailfile)                      , NULL),
 	DEF_OPT_STR("cwlogdir"                         , OFS(cwlogdir)                      , NULL),
@@ -563,14 +563,14 @@
 	DEF_OPT_INT32("httprefresh"             , OFS(http_refresh)             , 0),
 	DEF_OPT_INT32("httppollrefresh"         , OFS(poll_refresh)             , 60),
 	DEF_OPT_INT8("httphideidleclients"      , OFS(http_hide_idle_clients)   , 1),
-	DEF_OPT_STR("httphidetype"              , OFS(http_hide_type)           , NULL),
+	DEF_OPT_STR("httphidetype"              , OFS(http_hide_type)           , "sh"),
 	DEF_OPT_INT8("httpshowpicons"           , OFS(http_showpicons)          , 0),
 	DEF_OPT_INT8("httppiconsize"            , OFS(http_picon_size)          , 0),
 	DEF_OPT_INT8("httpshowmeminfo"          , OFS(http_showmeminfo)         , 0),
 	DEF_OPT_INT8("httpshowuserinfo"         , OFS(http_showuserinfo)        , 0),
 	DEF_OPT_INT8("httpshowreaderinfo"       , OFS(http_showreaderinfo)      , 0),
 	DEF_OPT_INT8("httpshowcacheexinfo"      , OFS(http_showcacheexinfo)     , 0),
-	DEF_OPT_INT8("httpshowecminfo"          , OFS(http_showecminfo)         , 0),
+	DEF_OPT_INT8("httpshowecminfo"          , OFS(http_showecminfo)         , 1),
 	DEF_OPT_INT8("httpshowloadinfo"         , OFS(http_showloadinfo)        , 0),
 	DEF_OPT_FUNC("httpallowed"              , OFS(http_allowed)             , iprange_fn, .free_value = iprange_free_fn),
 	DEF_OPT_INT8("httpreadonly"             , OFS(http_readonly)            , 0),
@@ -1253,6 +1253,29 @@
 
 static void dvbapi_boxtype_fn(const char *token, char *value, void *UNUSED(setting), FILE *f)
 {
+#if defined(__linux__)
+        char boxtype[128];
+        boxtype[0] = 0;
+
+	FILE *p = fopen("/proc/stb/info/boxtype", "r");
+	if (p) {
+		while (fgets(boxtype, sizeof(boxtype), p)) {
+			if (strstr(boxtype,"\n")) {
+				boxtype[strlen(boxtype)-1] = '\0';
+				break;
+			}
+		}
+		fclose(p);
+	}
+        if (boxtype[0]) {
+		if(!strcasecmp(boxtype, "sogno-8800hd") || !strcasecmp(boxtype, "unibox-hde")) {
+			value = "sogno";
+			cfg.dvbapi_boxtype = 14;
+			return;
+		}
+	}
+#endif
+
 	if(value)
 	{
 		int i;
@@ -1344,14 +1367,14 @@
 {
 	DEF_OPT_SAVE_FUNC(dvbapi_should_save_fn),
 	DEF_OPT_INT8("enabled"        , OFS(dvbapi_enabled)        , 0),
-	DEF_OPT_INT8("au"             , OFS(dvbapi_au)             , 0),
+	DEF_OPT_INT8("au"             , OFS(dvbapi_au)             , 1),
 	DEF_OPT_INT8("pmt_mode"       , OFS(dvbapi_pmtmode)        , 0),
 	DEF_OPT_INT8("request_mode"   , OFS(dvbapi_requestmode)    , 0),
 	DEF_OPT_INT32("listen_port"   , OFS(dvbapi_listenport)     , 0),
 	DEF_OPT_INT32("delayer"       , OFS(dvbapi_delayer)        , 0),
 	DEF_OPT_INT8("ecminfo_file"   , OFS(dvbapi_ecminfo_file)   , 1),
 	DEF_OPT_INT8("ecminfo_type"   , OFS(dvbapi_ecminfo_type)   , 0),
-	DEF_OPT_STR("user"            , OFS(dvbapi_usr)            , NULL),
+	DEF_OPT_STR("user"            , OFS(dvbapi_usr)            , "dvbapi"),
 	DEF_OPT_INT8("read_sdt"       , OFS(dvbapi_read_sdt)       , 0),
 	DEF_OPT_INT8("write_sdt_prov" , OFS(dvbapi_write_sdt_prov) , 0),
 	DEF_OPT_INT8("extended_cw_api", OFS(dvbapi_extended_cw_api), 0),
@@ -1483,7 +1506,7 @@
 		}
 #endif
 		NULLFREE(cfg.logfile);
-		cfg.logtostdout = 1;
+		cfg.logtostdout = 0;
 #ifdef HAVE_DVBAPI
 		cfg.dvbapi_enabled = 1;
 #endif
diff -Nur oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/oscam-time.c oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/oscam-time.c
--- oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/oscam-time.c	2022-12-30 18:32:42.483206787 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/oscam-time.c	2022-12-31 13:44:02.282047174 +0100
@@ -3,10 +3,6 @@
 
 static enum clock_type clock_type = CLOCK_TYPE_UNKNOWN;
 
-#if defined(CLOCKFIX)
-struct timeval lasttime; // holds previous time to detect systemtime adjustments due to eg transponder change on dvb receivers
-#endif
-
 int64_t comp_timeb(struct timeb *tpa, struct timeb *tpb)
 {
 	return (int64_t)(((int64_t)(tpa->time - tpb->time) * 1000ull) + ((int64_t) tpa->millitm - (int64_t) tpb->millitm));
@@ -131,18 +127,6 @@
 {
 	struct timeval tv;
 	gettimeofday(&tv, NULL);
-#if defined(CLOCKFIX)
-	if (tv.tv_sec > lasttime.tv_sec || (tv.tv_sec == lasttime.tv_sec && tv.tv_usec >= lasttime.tv_usec)) // check for time issues!
-	{
-		lasttime = tv; // register this valid time
-	}
-	else
-	{
-		tv = lasttime;
-		settimeofday(&tv, NULL); // set time back to last known valid time
-		//fprintf(stderr, "*** WARNING: BAD TIME AFFECTING WHOLE OSCAM ECM HANDLING, SYSTEMTIME SET TO LAST KNOWN VALID TIME **** \n");
-	}
-#endif
 	tp->time = tv.tv_sec;
 	tp->millitm = tv.tv_usec / 1000;
 }
@@ -151,18 +135,6 @@
 {
 	struct timeval tv;
 	gettimeofday(&tv, NULL);
-#if defined(CLOCKFIX)
-	if (tv.tv_sec > lasttime.tv_sec || (tv.tv_sec == lasttime.tv_sec && tv.tv_usec >= lasttime.tv_usec)) // check for time issues!
-	{
-		lasttime = tv; // register this valid time
-	}
-	else
-	{
-		tv = lasttime;
-		settimeofday(&tv, NULL); // set time back to last known valid time
-		//fprintf(stderr, "*** WARNING: BAD TIME AFFECTING WHOLE OSCAM ECM HANDLING, SYSTEMTIME SET TO LAST KNOWN VALID TIME **** \n");
-	}
-#endif
 	tp->time = tv.tv_sec;
 	tp->millitm = tv.tv_usec;
 }
@@ -281,11 +253,6 @@
 #define __GLIBCVER 9999
 #endif
 
-// Assume we have HAVE_pthread_condattr_setclock if CLOCK_MONOTONIC is defined
-#if defined(CLOCKFIX) && defined(CLOCK_MONOTONIC)
-#define HAVE_pthread_condattr_setclock 1
-#endif
-
 #if defined(HAVE_pthread_condattr_setclock)
 // UCLIBC 0.9.31 does not have pthread_condattr_setclock
 #if __UCLIBC_VER < 932
@@ -391,53 +358,8 @@
 {
 	struct timeval tv;
 	gettimeofday(&tv, NULL);
-#if defined(CLOCKFIX)
-	if (tv.tv_sec > lasttime.tv_sec || (tv.tv_sec == lasttime.tv_sec && tv.tv_usec >= lasttime.tv_usec)) // check for time issues!
-	{
-		lasttime = tv; // register this valid time
-	}
-	else
-	{
-		tv = lasttime;
-		settimeofday(&tv, NULL); // set time back to last known valid time
-		//fprintf(stderr, "*** WARNING: BAD TIME AFFECTING WHOLE OSCAM ECM HANDLING, SYSTEMTIME SET TO LAST KNOWN VALID TIME **** \n");
-	}
-#endif
-	ts->tv_sec = tv.tv_sec;
-	ts->tv_nsec = tv.tv_usec * 1000;
-	clock_type = CLOCK_TYPE_REALTIME;
-	return;
-#if 0
-#if !defined(CLOCKFIX) || (!defined(CLOCK_MONOTONIC) && !defined(__MACH__))
-	struct timeval tv;
-	gettimeofday(&tv, NULL);
 	ts->tv_sec = tv.tv_sec;
 	ts->tv_nsec = tv.tv_usec * 1000;
 	clock_type = CLOCK_TYPE_REALTIME;
 	return;
-#elif defined (__MACH__)
-// OS X does not have clock_gettime, use clock_get_time
-	clock_serv_t cclock;
-	mach_timespec_t mts;
-	host_get_clock_service(mach_host_self(), CALENDAR_CLOCK, &cclock);
-	clock_get_time(cclock, &mts);
-	mach_port_deallocate(mach_task_self(), cclock);
-	ts->tv_sec = mts.tv_sec;
-	ts->tv_nsec = mts.tv_nsec;
-	clock_type = CLOCK_TYPE_REALTIME;
-#else
-	if (clock_type == CLOCK_TYPE_REALTIME) // monotonic returned error
-	{
-		clock_gettime(CLOCK_REALTIME, ts);
-		return;
-	}
-	int32_t	ret = clock_gettime(CLOCK_MONOTONIC, ts);
-	clock_type = CLOCK_TYPE_MONOTONIC;
-	if ((ret < 0 && errno == EINVAL)) // Error fetching time from this source (Shouldn't happen on modern Linux)
-	{
-		clock_gettime(CLOCK_REALTIME, ts);
-		clock_type = CLOCK_TYPE_REALTIME;
-	}
-#endif
-#endif
 }
diff -Nur oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/README.config oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/README.config
--- oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/README.config	2022-12-30 18:44:46.461918175 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/README.config	2022-12-31 14:31:17.981645161 +0100
@@ -54,7 +54,7 @@
   ./config.sh -D card_readers -E CARDREADER_INTERNAL
 
 Available options:
-       addons: WEBIF WEBIF_LIVELOG WEBIF_JQUERY TOUCH WITH_SSL HAVE_DVBAPI IRDETO_GUESSING CS_ANTICASC WITH_DEBUG MODULE_MONITOR WITH_LB CS_CACHEEX CS_CACHEEX_AIO CW_CYCLE_CHECK LCDSUPPORT LEDSUPPORT CLOCKFIX IPV6SUPPORT
+       addons: WEBIF WEBIF_LIVELOG WEBIF_JQUERY TOUCH WITH_SSL HAVE_DVBAPI IRDETO_GUESSING CS_ANTICASC WITH_DEBUG MODULE_MONITOR WITH_LB CS_CACHEEX CS_CACHEEX_AIO CW_CYCLE_CHECK LCDSUPPORT LEDSUPPORT IPV6SUPPORT
     protocols: MODULE_CAMD33 MODULE_CAMD35 MODULE_CAMD35_TCP MODULE_NEWCAMD MODULE_CCCAM MODULE_CCCSHARE MODULE_GBOX MODULE_RADEGAST MODULE_SCAM MODULE_SERIAL MODULE_CONSTCW MODULE_PANDORA MODULE_GHTTP
       readers: READER_NAGRA READER_NAGRA_MERLIN READER_IRDETO READER_CONAX READER_CRYPTOWORKS READER_SECA READER_VIACCESS READER_VIDEOGUARD READER_DRE READER_TONGFANG READER_BULCRYPT READER_GRIFFIN READER_DGCRYPT
  card_readers: CARDREADER_PHOENIX CARDREADER_INTERNAL CARDREADER_SC8IN1 CARDREADER_MP35 CARDREADER_SMARGO CARDREADER_DB2COM CARDREADER_STAPI CARDREADER_STINGER
diff -Nur oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/README.md oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/README.md
--- oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/README.md	1970-01-01 01:00:00.000000000 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/README.md	2022-12-31 13:44:02.250047948 +0100
@@ -0,0 +1,15 @@
+[![Build Status](https://travis-ci.org/Schimmelreiter/oscam-smod.svg?branch=master)](https://travis-ci.org/Schimmelreiter/oscam-smod)
+
+oscam-smod
+==========
+ 
+This variant of oscam combines the benefits of
+- the original oscam-trunk (http://www.streamboard.tv/oscam/changeset)
+- Gorgone's modern interface (http://www.streamboard.tv/oscam-addons/changeset/1532/modern)
+and
+- oscam-emu (https://github.com/oscam-emu/oscam-patched/commits/master)
+- auto-configuration for many cards incl. HD+, Sky DE, TivuSat, ...  (WITH_CARDLIST)
+
+License
+-------
+Licensed under the GNU General Public License, Version 2.0. See [COPYING](https://github.com/Schimmelreiter/oscam-schimmelreiter/blob/master/src/COPYING) for more details.
diff -Nur oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/webif/include/css.css oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/webif/include/css.css
--- oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/webif/include/css.css	2022-12-30 18:47:59.557276215 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/webif/include/css.css	2022-12-31 13:46:42.518172448 +0100
@@ -733,7 +733,7 @@
 /* debugmenu	- status_sdebug.html
  * logmenu		- file.html
 */
-div.debugmenu, div.logmenu {
+div.debugmenu, div.logmenu, div.logsizemenu {
 	font-size:12px;
 	line-height:16px;
 	margin-bottom:8px
@@ -1908,4 +1908,3 @@
 	line-height:12px
 }
 #regex{margin-bottom: 20px; padding-bottom: 10px; margin-top: -10px;}
-#logsizemenu{margin-top: -25px; margin-bottom: 5px;}
\ Kein Zeilenumbruch am Dateiende.
diff -Nur oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/webif/status/status.html oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/webif/status/status.html
--- oscam-svn-modern-zeroconf-emu-icam-sogno-cak7/webif/status/status.html	2022-12-30 18:44:33.314234246 +0100
+++ oscam-svn-modern-zeroconf-emu-icam-sogno-cak7-smod/webif/status/status.html	2022-12-31 13:44:02.278047271 +0100
@@ -19,8 +19,8 @@
 </div>
 <DIV ID="subnav">
 	<UL ID="nav">
-		<LI CLASS="configmenu"><A HREF="http://www.streamboard.tv/oscam-addons/timeline" TARGET="_blank"><B>Oscam Modern ##CS_SVN_VERSION##</B></LI>
-		<LI CLASS="configmenu"><A HREF="https://trac.streamboard.tv/oscam/timeline" TARGET="_blank">Timeline</A></LI>
+		<LI CLASS="configmenu"><A HREF="https://github.com/Schimmelreiter/oscam-smod/commits/##CS_SMOD_VERSION_HASH##" TARGET="_blank"><B>##HTTPOSCAMLABEL## ##CS_VERSION## ##CS_SMOD_VERSION##</B></A></LI>
+		<LI CLASS="configmenu"><A HREF="https://trac.streamboard.tv/oscam/timeline" TARGET="_blank">(Trunk r##CS_SVN_VERSION##)</A></LI>
 		<LI CLASS="configmenu"><A HREF="#statusfooter">Status</A></LI>
 ##TPLPOLLINGSET##
 	</UL>
