Index: /trunk/module-webif-lib.h
===================================================================
--- /trunk/module-webif-lib.h	(revision 11520)
+++ /trunk/module-webif-lib.h	(revision 11521)
@@ -25,5 +25,5 @@
 #define AUTHNONCEHASHBUCKETS 4
 /* The maximum amount of GET parameters the webserver will parse. */
-#define MAXGETPARAMS 100
+#define MAXGETPARAMS 150
 /* The refresh delay (in seconds) when stopping OSCam via http. */
 #define SHUTDOWNREFRESH 30
